src/librustc/ty/subst.rs

```diff
@@ -298,20 +275,20 @@ impl<'a, 'gcx, 'tcx> Substs<'tcx> {
                        target_substs: &Substs<'tcx>)
                        -> &'tcx Substs<'tcx> {
         let defs = tcx.lookup_generics(source_ancestor);
-        Substs::new(tcx, target_substs.iter().chain(&self[defs.own_count()..]).cloned())
+        tcx.mk_substs(target_substs.iter().chain(&self[defs.own_count()..]).cloned())
```

Comment:
You can't avoid an iterator here. `.cloned()` is suspicious when used on a slice that can be passed directly - there's no cloning cost, these are just pointer copies.